### PR TITLE
Allow index creation in CREATE TABLE

### PR DESCRIPTION
**Pull request by treysp**
Teradata requires an index when a table is created (unless specifically overridden). It will create one automatically for a `CREATE TABLE [AS]` call, but people usually specify them explicitly. This PR allows those index specifications to be parsed.

I started by creating a new Teradata dialect, but since these changes are in the create parser/generator the dialect approach would have to contain independent copies of those. It seems like you're generally trying to avoid that, so I modified the core create parser/generator in this PR. Happy to change it to a dialect if that's your preference.

This PR implements the following changes:
- Allow `AND [NO] STATISTICS` option in `CREATE TABLE AS`
- Allow specification of index(es) in `CREATE TABLE` and `CREATE TABLE AS`
    - Support `UNIQUE` and `PRIMARY` index options
    - Support named or unnamed indexes
    - Support indexes with or without column specifications
    - Support arbitrarily many indexes

One implementation detail of note is adding a `parse_subquery_alias` argument to `_parse_select` that is passed on to `_parse_subquery` inside it. This argument prevents automatically parsing the token following a subquery as an alias.

### CODE REVIEW

File: sqlglot/generator.py [hunk 1]
```diff
@@ -434,6 +434,26 @@ def create_sql(self, expression: exp.Create) -> str:
...
+        no_primary_index = " NO PRIMARY INDEX" if expression.args.get("no_primary_index") else ""
+
+        indexes = expression.args.get("indexes")
+        index_sql = ""
+        if indexes is not None:
+            indexes_sql = []
+            for index in indexes:
+                ind_unique = " UNIQUE" if index.args.get("unique") else ""
+                ind_primary = " PRIMARY" if index.args.get("primary") else ""
+                ind_name = f' {index.args.get("this")}' if index.args.get("this") else ""
```
Comment (tobymao) [inline sqlglot/generator.py:453]: ind_name = f" {index.name}" if index.name else ""

File: sqlglot/parser.py [hunk 2]
```diff
@@ -1511,6 +1535,21 @@ def _parse_index(self) -> exp.Expression:
             columns=self._parse_expression(),
         )
 
+    def _parse_create_table_index(self):
+        unique = self._match_text_seq("UNIQUE")
```
Comment (tobymao) [inline sqlglot/parser.py:1539]: no need to do text because unique and primary are tokens

File: sqlglot/parser.py [hunk 3]
```diff
@@ -1511,6 +1535,21 @@ def _parse_index(self) -> exp.Expression:
...
+        primary = self._match_text_seq("PRIMARY")
+        self._match(TokenType.INDEX)
+        index = self._parse_id_var()
+        return self.expression(
+            exp.Index,
+            this=index,
+            columns=self.expression(
```
Comment (tobymao) [inline sqlglot/parser.py:1546]: i'm not sure we want tuples here, it should probably just be an array 

columns=self._parse_wrapped_csv(self._parse_column)

File: sqlglot/parser.py [hunk 4]
```diff
@@ -945,14 +948,32 @@ def _parse_create(self) -> t.Optional[exp.Expression]:
             this = self._parse_table(schema=True)
             properties = self._parse_properties()
             if self._match(TokenType.ALIAS):
-                expression = self._parse_select(nested=True)
+                expression = self._parse_select(nested=True, parse_subquery_alias=False)
```
Comment (tobymao) [inline sqlglot/parser.py:951]: why did you need to add this?

File: sqlglot/parser.py [hunk 5]
```diff
@@ -945,14 +948,32 @@ def _parse_create(self) -> t.Optional[exp.Expression]:
...
+                if self._match_text_seq("AND", "STATISTICS"):
+                    statistics = True
+                elif self._match_text_seq("AND", "NO", "STATISTICS"):
+                    statistics = False
+
+                no_primary_index = self._match_text_seq("NO", "PRIMARY", "INDEX")
+
+                indexes = []
+                while (
+                    self._match_text_seq("UNIQUE", advance=False)
```
Comment (tobymao) [inline sqlglot/parser.py:968]: use token

File: sqlglot/parser.py (hunk 2 above)
Comment (treysp) [reply sqlglot/parser.py:1539]: I'm not seeing a primary token (just primary key) - should I create one or use text?

File: sqlglot/parser.py (hunk 3 above)
Comment (treysp) [reply sqlglot/parser.py:1546]: Got it. Is this how I should handle it in the generator?

`ind_columns = f' ({", ".join([self.sql(col) for col in index.args.get("columns")])})'`

Comment (GeorgeSittas) [reply sqlglot/parser.py:1546]: > Got it. Is this how I should handle it in the generator?

You can use the `expressions` directly for this, I think. It defaults to `","` as a separator and you can provide `"columns"` as a key.

File: sqlglot/parser.py (hunk 4 above)
Comment (treysp) [reply sqlglot/parser.py:951]: This `_parse_select` calls `_parse_subquery` internally. If `parse_alias=False` isn't passed to `_parse_subquery` it assumes the token following the subquery is an alias. In this `CREATE TABLE` context, we know for sure that that token is not an alias.

For example, consider a command like `CREATE TABLE AS (SELECT a,b) INDEX (a)`. The token in the "INDEX" position will never be an alias for a subquery.

With default `_parse_subquery`:
image

With `parse_alias=False` passed down to `_parse_subquery`:
image

The latter is the output we're looking for.

Is there a better way to accomplish this?

Comment (tobymao) [reply sqlglot/parser.py:952]: i didn't understand what case this would be used for, make sure this is a test case

File: sqlglot/parser.py [hunk 6]
```diff
@@ -1512,6 +1533,22 @@ def _parse_index(self) -> exp.Expression:
             columns=self._parse_expression(),
         )
 
+    def _parse_create_table_index(self):
+        unique = self._prev.token_type == TokenType.UNIQUE
+        primary = self._prev.text.upper() == "PRIMARY" or self._match_text_seq("PRIMARY")
```
Comment (tobymao) [inline sqlglot/parser.py:1538]: why do you check _prev.text == primary and match_text primary?

File: sqlglot/parser.py [hunk 7]
```diff
@@ -946,14 +949,29 @@ def _parse_create(self) -> t.Optional[exp.Expression]:
...
 
+                if self._match_text_seq("AND", "STATISTICS"):
+                    statistics = True
+                elif self._match_text_seq("AND", "NO", "STATISTICS"):
+                    statistics = False
+
+                no_primary_index = self._match_text_seq("NO", "PRIMARY", "INDEX")
+
+                indexes = []
+                while (
```
Comment (tobymao) [inline sqlglot/parser.py:968]: maybe this should be

while (self._curr and self._curr.text.upper() in ("UNIQUE", "PRIMARY", "INDEX) 

or make UNIQUE, PRIMARY, INDEX in a constant,

then you won't need that code in the create_table_index section

that way you don't advance

Comment (tobymao) [reply sqlglot/parser.py:968]: do you need this condition at all? could you do

while True:
  self._parse_create_table_index()

   if not index:
      break

File: sqlglot/parser.py (hunk 6 above)
Comment (treysp) [reply sqlglot/parser.py:1538]: The text could be either "PRIMARY INDEX" or "UNIQUE PRIMARY INDEX", and the first word has already been consumed when we hit this line.

If "primary" is the first word in the text it will be _prev and "index" will be _curr. If primary is the second word in the text, "unique" will be _prev and "primary" will be _curr.

Comment (tobymao) [reply sqlglot/parser.py:1538]: then you can do

self._match_text_seq("PRIMARY", "INDEX") 
self._match_text_seq("UNIQUE", "PRIMARY", "INDEX")

File: sqlglot/parser.py (hunk 4 above)
Comment (treysp) [reply sqlglot/parser.py:952]: Is this an appropriate test case in test_parser.py `test_select`?

```
self.assertIsNone( 
    parse_one("create table a as (select b from c) index").find(exp.TableAlias)
    )
```

File: sqlglot/parser.py (hunk 6 above)
Comment (treysp) [reply sqlglot/parser.py]: Sorry, I worded that comment poorly - there are actually 6 distinct text sequences

File: sqlglot/parser.py (hunk 7 above)
Comment (treysp) [reply sqlglot/parser.py]: Yes, that's a much better solution!

File: sqlglot/parser.py (hunk 4 above)
Comment (tobymao) [reply sqlglot/parser.py:952]: ok

**Review by GeorgeSittas — APPROVED** [review]
LGTM

**Review by tobymao — APPROVED** [review]
